Problems following instructions to build Kernel

I carefully followed https://www.raspberrypi.com/documentation/computers/linux_kernel.html#building-the-kernel-locally, to build a 32-bit kernel from scratch for use in developing a device driver for my RPi400. However, building the kernel  (make -j4 zImage modules dtbs) failed very quickly, complaining of an unknown target "zImage".

I eventually found the cause to be that while KERNEL was correctly defined as a shell variable, it was not being exported, so make did not see it in its environment.  The fix is simply to prefix the KERNEL= definition with "export ".

I have created a branch with the relevant lines corrected.

@pelwell  I did follow the instructions to the letter, and I did experience the problem as reported. However my explanation of what was wrong and my proposed solution are not correct. 

I now believe the problem's root cause is related to what @lurch mentioned, and involves an unexpected (and to my mind, undesirable) dependency of the local kernel build process on the version of the kernel that the build is run on. Here's a summary of the history:

I was using a Pi 400 running a freshly downloaded 32b Raspberry Pi OS. None of the instructions that I had followed mentioned that the default kernel running in this case is actually 64b, despite the system being "32b", so I was unaware of that. 

I wanted to build a 32b kernel to run on the same machine and also a 2GB Pi 4B (the target for my current development), neither of which need a 64b kernel. So I followed the instructions which say to set KERNEL=kernel7l before doing "make bcm2711_defconfig". I then customised my build version in CONFIG_LOCALVERSION - though I was puzzled why it had "v8" in the default value - by suffixing it with "-MT". When I then I typed "make -j4 zImage", the output was (just as I reported above):

```
  SYNC    include/config/auto.conf.cmd
make: *** No rule to make target 'zImage'.  Stop.
```

Subsequently, I found out about the 32b vs 64b kernel version detail that @lurch mentioned. So I switched to working under a 32b kernel (arm_64bit=0 in /boot/config.txt).  But being still puzzled about the "-v8" version I had seen for CONFIG_LOCALVERSION on the first attempt, I prefixed "KERNEL=kernel7l" with "export" on the assumption that that might fix the problem.  After running make bcm2711_defconfig and opening .config to customise it, the existing value was now "-v7l", so I supposed that the "export" had fixed the problem. That the kernel build also then succeeded further strengthened this conclusion, and hence I created a patch and submitted this issue / pull request.

However, after reading your comment above, further digging into the matter has revealed that, despite all appearances from the instructions, the setting of KERNEL has no effect at all on what kernel version is actually built! Rather, it seems the version is determined purely from the version of the kernel that the build is run **on**. 

Ideally (as I see it), the makery should be improved to allow building any target kernel version, independent of the kernel version the build is run under. But if not, then the instructions need an update to make the matter clear (you can only do a local build of an 32-bit kernel under a 32-bit kernel, and the same for 64-bit), or to make explicit that you must do such a build as a cross-compilation, not a "local" one, despite the host and target machines being the same. 

I hope that all makes sense?

--- a/documentation/asciidoc/computers/linux_kernel/building.adoc
+++ b/documentation/asciidoc/computers/linux_kernel/building.adoc
@@ -12,7 +12,7 @@ The instructions below are divided into native builds and cross-compilation; cho
 
 === Building the Kernel Locally
 
-IMPORTANT: If you are build a 32-bit kernel on the 32-bit distribution of Raspberry Pi OS you should be aware that while the user land in this distribution is 32-bit, the kernel itself is a 64-bit kernel. You therefore need to follow the <<cross-compiling-the-kernel,instructions for cross-compiling the kernel>>.  
+IMPORTANT: Building the 64-bit kernel on the 32-bit distribution of Raspberry Pi OS is a cross-compilation exercise because it requires the installation of the cross-compiler (`gcc-aarch64-linux-gnu`). Building the 32-bit kernel on the 32-bit distribution of Raspberry Pi OS — which has a 32-bit userland, and 64-bit kernel — you should set `ARCH=arm`, although it's likely that this may already be set as `arm_64bit=0` will have been set. Instructions for <<cross-compiling-the-kernel,cross-compiling the kernel>> can be found later on this page.  
 
 On a Raspberry Pi, first install the latest version of https://www.raspberrypi.com/software/operating-systems/#raspberry-pi-os-32-bit[Raspberry Pi OS]. Then boot your Raspberry Pi, log in, and ensure you're connected to the internet to give you access to the sources.
 

--- a/documentation/asciidoc/computers/linux_kernel/building.adoc
+++ b/documentation/asciidoc/computers/linux_kernel/building.adoc
@@ -12,6 +12,8 @@ The instructions below are divided into native builds and cross-compilation; cho
 
 === Building the Kernel Locally
 
+IMPORTANT: If you are build a 32-bit kernel on the 32-bit distribution of Raspberry Pi OS you should be aware that while the user land in this distribution is 32-bit, the kernel itself is a 64-bit kernel. You therefore need to follow the <<cross-compiling-the-kernel,instructions for cross-compiling the kernel>>.  
+
 On a Raspberry Pi, first install the latest version of https://www.raspberrypi.com/software/operating-systems/#raspberry-pi-os-32-bit[Raspberry Pi OS]. Then boot your Raspberry Pi, log in, and ensure you're connected to the internet to give you access to the sources.
 
 First install Git and the build dependencies:

